[Bug]~new pnpm v10 version has conflicts with bcrypt 5.1.1

### Verify latest release

- [ ] I verified that the issue exists in the latest pnpm release

### pnpm version

_No response_

### Which area(s) of pnpm are affected? (leave empty if unsure)

_No response_

### Link to the code that reproduces this issue or a replay of the bug

_No response_

### Reproduction steps

Error: Cannot find module '/usr/src/app/node_modules/.pnpm/bcrypt@5.1.1/node_modules/bcrypt/lib/binding/napi-v3/bcrypt_lib.node'

### Describe the Bug

Error: Cannot find module '/usr/src/app/node_modules/.pnpm/bcrypt@5.1.1/node_modules/bcrypt/lib/binding/napi-v3/bcrypt_lib.node'

### Expected Behavior

there should not be any error

### Which Node.js version are you using?

node-20 slim

### Which operating systems have you used?

- [ ] macOS
- [ ] Windows
- [x] Linux

### If your OS is a Linux based, which one it is? (Include the version if relevant)

_No response_

--- a/.changeset/proud-crabs-own.md
+++ b/.changeset/proud-crabs-own.md
@@ -1,0 +1,8 @@
+---
+"@pnpm/headless": patch
+"@pnpm/build-modules": patch
+"@pnpm/core": patch
+"pnpm": patch
+---
+
+Don't read a package from side-effects cache if it isn't allowed to be built [#9042](https://github.com/pnpm/pnpm/issues/9042).

--- a/pkg-manager/core/src/install/link.ts
+++ b/pkg-manager/core/src/install/link.ts
@@ -395,6 +395,7 @@ async function linkNewPackages (
         optional: opts.optional,
       }),
     linkAllPkgs(opts.storeController, newPkgs, {
+      allowBuild: opts.allowBuild,
       depGraph,
       depsStateCache: opts.depsStateCache,
       disableRelinkLocalDirDeps: opts.disableRelinkLocalDirDeps,

